Marlin/cardreader.cpp

```diff
@@ -577,9 +633,187 @@ void CardReader::updir() {
+ *  - Most RAM: Buffer the directory and return filenames from RAM
+ */
+void CardReader::presort()
+{
+  flush_presort();
+
+  uint16_t fileCnt = getnrfilenames();
+  if (fileCnt > 0) {
+
+    if (fileCnt > SDSORT_LIMIT) fileCnt = SDSORT_LIMIT;
+
+    #if ENABLED(SDSORT_USES_RAM)
+      #if ENABLED(SDSORT_USES_MORE_RAM)
+        sortshort = new char*[fileCnt]; // keep in the heap
+        sortnames = new char*[fileCnt]; // keep in the heap
+        #if FOLDER_SORTING || ENABLED(SDSORT_GCODE)
+          // More RAM option pre-allocates
+          #if ENABLED(SDSORT_USES_MALLOC)
+            isDir = (uint8_t*)malloc((fileCnt+7)>>3); // in the heap
+          #endif
+        #endif
+      #else
+        char *sortnames[fileCnt]; // on the stack
+        #if FOLDER_SORTING || ENABLED(SDSORT_GCODE)
+          uint8_t isDir[(fileCnt+7)>>3]; // on the stack
+        #endif
+      #endif
+
+      sort_order = new uint8_t[fileCnt];
+
+    #else
+      char name1[LONG_FILENAME_LENGTH+1]; // on the stack
+    #endif
+
+    if (fileCnt > 1) {
+
+      // Init sort order. If using RAM then read all filenames now.
+      for (uint16_t i = 0; i < fileCnt; i++) {
+        sort_order[i] = i;
+        #if ENABLED(SDSORT_USES_RAM)
+          getfilename(i);
+          sortnames[i] = strdup(LONGEST_FILENAME); // malloc
+          #if ENABLED(SDSORT_USES_MORE_RAM)
+            sortshort[i] = strdup(filename); // malloc
+          #endif
+          // char out[30];
+          // sprintf_P(out, PSTR("---- %i %s %s"), i, filenameIsDir ? "D" : " ", sortnames[i]);
+          // SERIAL_ECHOLN(out);
+          #if FOLDER_SORTING || ENABLED(SDSORT_GCODE)
+            uint16_t mod = i & 0x07, ind = i >> 3;
+            if (mod == 0) isDir[ind] = 0x00;
+            if (filenameIsDir) isDir[ind] |= _BV(mod);
+          #endif
+        #endif
+      }
+
+      // Bubble Sort
+      #if ENABLED(SDSORT_GCODE)
+        if (sort_alpha)
+      #endif
+      for (uint16_t i = fileCnt; --i;) {
+        bool cmp, didSwap = false;
+        for (uint16_t j = 0; j < i; ++j) {
+          uint16_t s1 = j, s2 = j + 1, o1 = sort_order[s1], o2 = sort_order[s2];
+
+          #if FOLDER_SORTING || ENABLED(SDSORT_GCODE)
+            #if ENABLED(SDSORT_USES_RAM)
+              uint8_t ind1 = o1 >> 3, mod1 = o1 & 0x07, mask1 = 1 << mod1,
+                      ind2 = o2 >> 3, mod2 = o2 & 0x07, mask2 = 1 << mod2;
+              #define _SORT_CMP_NODIR() (strcasecmp(sortnames[o1], sortnames[o2]) > 0)
+              #define _SORT_CMP_DIR(fs) \
+                (((isDir[ind1] & mask1) != 0) == ((isDir[ind2] & mask2) != 0) \
+                  ? _SORT_CMP_NODIR \
+                  : (isDir[fs > 0 ? ind1 : ind2] & (fs > 0 ? mask1 : mask2)) != 0)
+            #else
+              #define _SORT_CMP_NODIR() (strcasecmp(name1, name2) > 0)
+              #define _SORT_CMP_DIR(fs) ((dir1 == filenameIsDir) ? _SORT_CMP_NODIR : (fs > 0 ? dir1 : !dir1))
+            #endif
```

Comment:
```
-+              #define _SORT_CMP_DIR(fs) ((dir1 == filenameIsDir) ? _SORT_CMP_NODIR : (fs > 0 ? dir1 : !dir1))
++              #define _SORT_CMP_DIR(fs) ((dir1 == filenameIsDir) ? _SORT_CMP_NODIR() : (fs > 0 ? dir1 : !dir1))
```
else
cardreader.cpp:741: error: '_SORT_CMP_NODIR' was not declared in this scope